### PR TITLE
[move-smith] Add a basic Move code generator and the fuzzing pipeline

### PR DESCRIPTION
**Pull request by zzjas**
## Description

Adding the very basic setup for fuzzing Move compiler and VM at the source code level.

The simple code generator follows the idea of [WasmSmith](https://fitzgeraldnick.com/2020/08/24/writing-a-test-case-generator.html). The current implementation keeps tracks of scoping and typing information and is able to generate simple functions, numerical literals, bindings, and function calls.

The fuzzing pipeline is functional as well as the coverage reports. However, the current fuzz target only tries to compile the generated module with compiler V1 (I believe, the code is [here](https://github.com/zzjas/aptos-core/blob/move-smith-init/third_party/move/tools/move-smith/src/utils.rs#L16-L35)).

The `README.md` has more detailed information for usage and the code structure.

## Type of Change
- [x] New feature

## Which Components or Systems Does This Change Impact?
- [x] Move/Aptos Virtual Machine

## How Has This Been Tested?

Through the two simple integration tests and manually running `cargo fuzz` and `./scripts/check_output.sh`.

## Key Areas to Review

* The overall setup
* The Rust coding practice
* The AST, tho it's mostly taken from the [parse AST](https://github.com/aptos-labs/aptos-core/blob/main/third_party/move/move-compiler/src/parser/ast.rs)

## Checklist
- [x] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I identified and added all stakeholders and component owners affected by this change as reviewers
- [x] I tested both happy and unhappy path of the functionality
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

File: third_party/move/tools/move-smith/README.md
Comment (vineethk) [inline third_party/move/tools/move-smith/README.md:16]: Could you add a bit more context on fuzz targets, and maybe a small sentence on each available option?

File: third_party/move/tools/move-smith/src/ast.rs
Comment (vineethk) [inline third_party/move/tools/move-smith/src/ast.rs:9]: All public structs, functions, methods, enums, etc should have at least a brief documentation. They should list any invariants to be maintained on datastructures, pre-conditions and post-conditions if applicable.

File: third_party/move/tools/move-smith/src/ast.rs
Comment (vineethk) [inline third_party/move/tools/move-smith/src/ast.rs:22]: I'm ignoring the commented out code for now, but in general, we should not leave commented out code.

File: third_party/move/tools/move-smith/src/ast.rs
Comment (vineethk) [inline third_party/move/tools/move-smith/src/ast.rs:3]: Each module should have some high-level documentation, guiding a developer. `See third_party/move/move-compiler-v2/src/pipeline/unreachable_code_analysis.rs` as an example.

File: third_party/move/tools/move-smith/src/codegen.rs
Comment (vineethk) [inline third_party/move/tools/move-smith/src/codegen.rs:7]: I wonder if we could just use the `std::fmt::Display` trait here, which automatically implements the `ToString` trait?

File: third_party/move/tools/move-smith/src/names.rs
Comment (vineethk) [inline third_party/move/tools/move-smith/src/names.rs:9]: Could we use newtype pattern here instead (https://doc.rust-lang.org/rust-by-example/generics/new_types.html) to get additional type safety benefits instead of just type name aliasing?

File: third_party/move/tools/move-smith/src/names.rs
Comment (vineethk) [inline third_party/move/tools/move-smith/src/names.rs:10]: Needs documentation about the semantics and invariants of this data structure.

File: third_party/move/tools/move-smith/scripts/check_output.sh
Comment (vineethk) [inline third_party/move/tools/move-smith/scripts/check_output.sh:2]: Add brief comment note on what this script does.

File: third_party/move/tools/move-smith/scripts/coverage.sh
Comment (vineethk) [inline third_party/move/tools/move-smith/scripts/coverage.sh:2]: Same comment as `check_output.sh`.

**Review by vineethk — APPROVED** [review]
Generally looks pretty good - I have some minor comments to consider. Ok to merge after addressing those comments.

File: third_party/move/tools/move-smith/src/codegen.rs
Comment (zzjas) [reply third_party/move/tools/move-smith/src/codegen.rs:7]: The main reason I created this trait is to have the `emit_code_lines` that returns `Vec<String>`. This allows me to easily add spaces for indentation before each line. If only a `String` is returned, then I'll need to split, pad, and concat. I don't know if `std::fmt::Display` would also give this flexibility.

Comment (vineethk) [reply third_party/move/tools/move-smith/src/codegen.rs:7]: Gotcha. `Display` would not give you that flexibility. Let's keep it as is then.

File: third_party/move/tools/move-smith/README.md
Comment (zzjas) [reply third_party/move/tools/move-smith/README.md]: Done

File: third_party/move/tools/move-smith/src/ast.rs
Comment (zzjas) [reply third_party/move/tools/move-smith/src/ast.rs:22]: Will keep this in mind and make sure no commented out code before merging into main

File: third_party/move/tools/move-smith/scripts/coverage.sh
Comment (zzjas) [reply third_party/move/tools/move-smith/scripts/coverage.sh:2]: Done

File: third_party/move/tools/move-smith/src/names.rs
Comment (zzjas) [reply third_party/move/tools/move-smith/src/names.rs]: Done

File: third_party/move/tools/move-smith/scripts/coverage.sh
Comment (zi0Black) [inline third_party/move/tools/move-smith/scripts/coverage.sh:46]: Remove all hard-coded paths:
```
# Get rust lib bin path to access llvm-cov in nightly toolchain
RUSTUP_TOOLCHAIN=$(rustup toolchain list | grep nightly | awk '{print $1}')
# default host triple
DHT=$(rustup run $RUSTUP_TOOLCHAIN rustc -vV | grep host | awk '{print $2}')
RUST_LIB_BIN_PATH="$(rustup run $RUSTUP_TOOLCHAIN rustc --print sysroot)/lib/rustlib/$DHT/bin"
```
or use `find`:
```
binary=$(find ./target/*/coverage/ -name $fuzz_target -type f -perm /111)
```

File: third_party/move/tools/move-smith/src/config.rs
Comment (zi0Black) [inline third_party/move/tools/move-smith/src/config.rs:21]: What criteria were used to pick those numbers?

Comment (zzjas) [reply third_party/move/tools/move-smith/src/config.rs:21]: They were just randomly chosen for now.

File: third_party/move/tools/move-smith/src/ast.rs
Comment (zzjas) [reply third_party/move/tools/move-smith/src/ast.rs:17]: Added

File: third_party/move/tools/move-smith/src/ast.rs
Comment (zzjas) [reply third_party/move/tools/move-smith/src/ast.rs:3]: Added

File: third_party/move/tools/move-smith/scripts/coverage.sh
Comment (zzjas) [reply third_party/move/tools/move-smith/scripts/coverage.sh]: Updated with `find`.

**Comment by zzjas** [comment]
Hi @vineethk @zi0Black thanks for the comments! I believe all of them have been addressed. Please take another look!

File: third_party/move/tools/move-smith/src/ast.rs
Comment (vineethk) [inline third_party/move/tools/move-smith/src/ast.rs:61]: I am assuming we are limiting to at most one return statement, only to get started with and not as limitation.

File: third_party/move/tools/move-smith/src/move_smith.rs
Comment (vineethk) [inline third_party/move/tools/move-smith/src/move_smith.rs:396]:
```suggestion
    fn generate_declaration(
```

**Review by vineethk — APPROVED** [review]

File: third_party/move/tools/move-smith/src/ast.rs
Comment (zzjas) [reply third_party/move/tools/move-smith/src/ast.rs:61]: Good catch! Yes currently only one return will be generated but we could easily generate more returns later. Will keep this point in mind but will add the functionality in future PRs.

File: third_party/move/tools/move-smith/src/move_smith.rs
Comment (zzjas) [reply third_party/move/tools/move-smith/src/move_smith.rs]: Fixed!

**Pull request merged**
